### PR TITLE
Reduce contention for Datacontract Serialization

### PR DESCRIPTION
**Pull request by Daniel-Svensson**
Making profiling/stress testing my rewrite of OpenRiaServices from WCF to aspnet core I identified a bottleneck in the datacontract serializer. 

I detected some contention for an API returning around 20 000 objects of a [fairly simple type](https://github.com/OpenRIAServices/Benchmarks/blob/9b39102949b8dcfd79054779cc431443608a72ae/OpenRiaServices.Client.Benchmarks/Server/Cities/CityTypes.cs#L48) with a datacontract surrogate which looks the same.

This fix is made in 2 steps, 
- Improve GetBuiltInDataContract performance
- Improve GetID performance

Loadtesting was done from local machine so the load testing (Netling/Westwind Web Surge) also consume resources 

Initially GetBuiltInDataContract was the source of contention.
Contention time for 20s run:
image
* The lock + Dictionary was replaced by a ConcurrentDictionary instead removing the contention.

When fixing that most contention seemed to move to GetId:
Contention time for 20s run after first commit:
image

* The lock + Dictionary was replaced by a ConcurrentDictionary and the lock for the cache is only taken on update

### Other possible solutions / comments
- The first fix maybe setup the whole dictionary on first access and then keeping the dictionary read only ?
  I decided to keep current logic in order to  keep the change small and not affect startup/first access performance to much,
- I looked into caching a static Func for the factory method, but I got problem due to the RequiresUnreferencedCode so i used a static lambda instead (which seems to be cached by roslyn, but with an extra if statement)

### Numbers
On 16 AMD 5800X

### 8 thead loading
| |  net7 preview 4 |  net7 preview 4  + first commit |  net7 preview 4 + second commits|net7 preview 4 + 3rd commits
|---|--|--|--|--|
|req/sec| 115 req/sec |  235 req/sec |  255 req/sec|258 req/sec|
|data served|  5.1gb |  10.5gb |  11,3gb| 11,5gb|
|95%th|82,7ms|47,1ms|44,7ms|42,43ms|
|99%th|94,9ms|72ms|63,6ms|51,51ms|

### 16 threads loading

From left to right: 
- net7 preview 4 "original" (with r2r)
- net7 preview 4  + first commit (~80% cpu load, including load generation)
- ~net7 preview 4 + second commits~ (100% cpu load including load generation)
image
- net7 preview 4 + 3rd commit (lazy) (100% cpu load)

* net7 preview 4 + 4th commit (lazy + changed key)
* 

* net7 preview 6 + 7th commit (changed key + int value and TryGet similar to code in https://github.com/dotnet/runtime/pull/71752)
* 

### 1  or 2 threads loaded
~~No significant difference between the versions to  draw any conclusions.~~
**update August**:  performance seems to be better even with no or little concurrency.
* ~5% for 1 thread (78 -> 82 rps)
* ~16% for 2 threads (135->157 rps)

### Other

* When only serializing 10 objects per query the improvement drops there is still a measureable improvement ~165 000 rps  vs ~<143 000 rps for 16 threads (console logging set to warning)

### CODE REVIEW

File: src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs [hunk 1]
```diff
@@ -441,36 +449,34 @@ private static bool ContractMatches(DataContract contract, DataContract cachedCo
...
                 {
-                    IntRef? id;
-                    typeHandle = GetDataContractAdapterTypeHandle(typeHandle);
-                    s_typeHandleRef.Value = typeHandle;
-                    if (!s_typeToIDCache.TryGetValue(s_typeHandleRef, out id))
+                    var id = s_typeToIDCache.GetOrAdd(typeHandleRef, static (key) =>
                     {
-                        int value = s_dataContractID++;
-                        if (value >= s_dataContractCache.Length)
+                        lock (s_cacheLock)
```
Comment (drieseng) [inline src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs:458]: Just an observation: under load, you'll create unused entries in **s_dataContractCache** as **ConcurrentDictionary<TKey,TValue>** does not guarantee that the specified delegate is only invoked once for a given key.

Comment (Daniel-Svensson) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs:458]: That is a good observation.
It think something should be done about that commit.

I wonder if it makes sense to use Lazy to protect against that or do something similar to "double checked locking" (using TryGet before lock) or just use a reader/writer lock and fall back to dictionary. 

I am open to drop that commit for know and maybe open a separate pr at a later time after comparing the alternative fixes

Comment (Daniel-Svensson) [reply src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs]: I did a small benchmarks (single threaded), which just invokes GetId for different keys with almost no adds just to measure overhead for the non-contented case and settled for using concurrent dictionary + lazy.

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 7 5800X, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100-preview.4.22252.9
  [Host]     : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT
  DefaultJob : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT


```
|     Method |     Mean |     Error |    StdDev |
|----------- |---------:|----------:|----------:|
|   Original | 2.380 μs | 0.0126 μs | 0.0118 μs |
|       Lazy + Concurrent (third commit)| 3.076 μs | 0.0208 μs | 0.0195 μs |
|     Faulty (second commit)  | 3.111 μs | 0.0369 μs | 0.0327 μs |
| ReaderWriterLockSlim and dictionary| 5.021 μs | 0.0192 μs | 0.0180 μs |

**Comment by Daniel-Svensson** [comment]
Commenting reply here as well since it is hidden when resolved.

I did a small benchmarks (single threaded), which just invokes GetId for different keys with almost no adds just to measure overhead for the non-contented case and settled for using concurrent dictionary + lazy.

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 7 5800X, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100-preview.4.22252.9
  [Host]     : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT
  DefaultJob : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT


```
|     Method |     Mean |     Error |    StdDev |
|----------- |---------:|----------:|----------:|
|   Original | 2.380 μs | 0.0126 μs | 0.0118 μs |
|       Lazy + Concurrent (third commit)| 3.076 μs | 0.0208 μs | 0.0195 μs |
|     Faulty (second commit)  | 3.111 μs | 0.0369 μs | 0.0327 μs |
| ReaderWriterLockSlim and dictionary| 5.021 μs | 0.0192 μs | 0.0180 μs |


In original code GetBuiltInDataContract took 9,8% of time GetId took 0.5% of time.
After the numbers are 0,3% and 0,08%

**Comment by StephenMolloy** [comment]
We have a very large PR in the works trying to reconcile the strange port of DCS that has been in .Net Core with the more complete version in .Net 4.8. (#71752). In doing this reconciliation, I believe this issue is also addressed - in a similar but not exactly the same manner. After that PR goes through though, you should see performance improvement here. If not, please reopen the issue.

**Comment by Daniel-Svensson** [comment]
@StephenMolloy do you have time frame for your large pr, Is there any chance of this getting reviewed in time to make it possible for a merge before the net7 cut-off date at August 15 ?

I did a quick test if your pr and it only improves performance marginally so I think it still makes sense to improve the 2 methods that are changed here.
FB924B45071F42D59E3C6E2B51073905

**Comment by StephenMolloy** [comment]
As mentioned earlier, a similar approach on this particular issue has been brought back from .Net 4.8 in our mega reconciliation of 7.0 with 4.8. #71752 has been checked in now. Hopefully you see this show up in your performance testing of RC1.

**Pull request closed without merging**